### PR TITLE
fix: add missing vLLM TP allreduce in DeepSeekModel context+generation ops

### PR DESCRIPTION
**Pull request by simone-chen**
#### Overview:

`DeepSeekModel` on the vLLM backend was emitting **0 ms** for the TP allreduces that run after attention output projection and after MoE in every transformer layer. Silicon (Kimi-K2.5 8k1k FP4 at TP=4) spends ~40 ms / mixed step on them. This PR models the cost in both `context_ops` and `generation_ops`, for both the KIMIK25 and DEEPSEEK families.

#### Details:

Three commits:

- **baad0fc4** `generation_tp_allreduce`: one `ops.CustomAllReduce` appended to `DeepSeekModel.generation_ops`, scale = `2 × num_layers × mtp_scale_factor`, gated on `self._backend_name == "vllm"`.
- **2df8328b** `context_tp_allreduce`: sibling op appended to `DeepSeekModel.context_ops`, scale = `2 × num_layers` (context_ops is not MTP-scaled, matching the rest of the prefill ops).
- **b8957add** apply to DEEPSEEK family too: thread `backend_name` through `DeepSeekModel.create()`'s DEEPSEEK fallback so DS-V3 / R1 + vLLM also gets the AR ops (and the existing vLLM standard-attention swap that was previously dormant for non-KIMIK25). Also extract `v_head_dim` from `DeepSeekForCausalLM` / `DeepseekV3ForCausalLM` config in the parser — same MLA architecture as Kimi K2.5, but the parser was only extracting v_head_dim for the KIMIK25 family, causing `_vllm_head_size` to silently fall back to `hidden_size // n_heads = 56` instead of the real 128.

Both AR ops reuse `ops.CustomAllReduce` → `PerfDatabase.query_custom_allreduce`, which already loads the vLLM-native `cross_device_reduce` rows produced by `collector/collect_all_reduce.py` (`kernel_source="vLLM_custom_graph"`). No data-collection or loader changes.

Per-layer cost (1× attention AR + 1× MoE AR) matches the standard vLLM forward shape; the shared expert's `down_proj` uses `reduce_results=False` so routed+shared are summed before a single combined AR fires at MoE exit, *not* two separate ARs.

##### Source-level verification (vLLM main)

Cross-checked silicon kernel counts against vLLM source.

1. **Attention `o_proj` → AllReduce** ✅
   `vllm/model_executor/models/deepseek_v2.py:861`, `DeepseekV2MLAAttention`:
   ```python
   self.o_proj = RowParallelLinear(
       self.num_heads * self.v_head_dim,
       self.hidden_size,
       bias=False, ...
   )
   ```
   `RowParallelLinear` defaults to `reduce_results=True` → calls `tensor_model_parallel_all_reduce` at end of forward. One AR per layer.

2. **MoE → AllReduce** ✅
   `vllm/model_executor/layers/fused_moe/runner/moe_runner.py:~491`, `_maybe_reduce_final_output`:
   ```python
   if (
       not self.moe_config.is_sequence_parallel
       and (self.moe_config.tp_size > 1 or self.moe_config.ep_size > 1)
       and not self._fused_output_is_reduced
   ):
       states = tensor_model_parallel_all_reduce(states)
   ```
   Fires when `tp_size > 1 OR ep_size > 1`, sequence-parallel is off, and the combine kernel didn't pre-reduce.

3. **Shared expert** uses `reduce_results=False`, confirming the fused single-AR pattern at MoE exit rather than two separate ARs.

4. **LM head** uses `ParallelLMHead` → `RowParallelLinear` → 1 AR per forward pass.

##### Tally

- vLLM source (Kimi-K2.5 / DeepSeek V3, 61 layers): 2 × 61 + 1 = **123** AllReduces per forward pass.
- Silicon nsys per mixed step: **123× `ncclDevKernel_AllReduce_Sum_bf16_RING_LL`**.
- AIC modeled here: **2 × num_layers = 122** per phase. The +1 logits AR is intentionally not modeled (~0.8% of cost).

##### Expected impact

- Closes the **+38 ms / mixed-step** TTFT gap reported for Kimi-K2.5 8k1k FP4 TP=4 (context_ops side, 2df8328b).
- Adds **+0.7 ms / decode-step** (generation_ops side, baad0fc4). Unfused-AR proxy for the FlashInfer fused AllReduce+RMSNorm kernel that fires only in pure-decode CUDA-graph steps; separate fused modeling is a follow-up.
- Lifts TTFT accuracy for **all TP>1 vLLM workloads** on Kimi K2.5 AND DeepSeek V3 / R1 (the latter via b8957add).
- Scope is limited to the `DeepSeekModel` class. `TrtllmWideEPDeepSeekModel` and `WideEPDeepSeekModel` already model their respective collectives via their WideEP / NCCL ops.

##### Source-verified for other backends (not applied here)

- **TRT-LLM** narrow-TP DeepSeek (`tensorrt_llm/_torch/models/modeling_deepseekv3.py`): same shape — per-layer AR after attention (`MLA` with `reduce_output=True` → RowLinear `o_proj`) and after MoE (`self.allreduce(...)` / `MoEAllReduce`). Both skip when `enable_attention_dp=True`. Could be widened to "trtllm" in a follow-up.
- **SGLang** narrow-TP DeepSeek (`python/sglang/srt/models/deepseek_v2.py`): different shape — attention `o_proj` instantiated with `reduce_results=False`, attention-side reduction relocated to a layer-level communicator (SP-style reduce_scatter/all_gather). Naively widening the AR gate would mis-model the cost; needs separate op layout.

##### Out of scope

- Fused FlashInfer AllReduce+RMSNorm for pure-decode (AllReduceFusionPass).
- The +1 logits AR per forward pass (~0.8% of AR cost).
- TRT-LLM and SGLang DeepSeek narrow-TP AR modeling (separate follow-up tickets).
- `is_sequence_parallel` / `_fused_output_is_reduced` MoE-AR gates (would require new model config plumbing).

#### Where should the reviewer start?

- `src/aiconfigurator/sdk/models/deepseek.py` — the two new `ops.CustomAllReduce` appendings (one in context, one in generation, both gated on `self._backend_name == "vllm"`), and the one-line `backend_name=backend_name` thread-through in `DeepSeekModel.create()`.
- `src/aiconfigurator/sdk/utils.py:588-596` — `v_head_dim` extraction added to DEEPSEEK V3 architecture's `extra_params`. Same key Kimi K2.5 already uses; bug-fix-style change.
- `tests/unit/sdk/models/test_model_config.py::TestDeepSeekTPAllReduce` — six cases: vLLM (Kimi K2.5) emits both AR ops with correct scale, vLLM (DS-V3) emits both ARs AND has `_vllm_head_size == 128`, vLLM at TP=1 still carries the ops (no-op via the op's own short-circuit), trtllm/sglang narrow-EP paths don't emit either.

#### Tests

```
$ .venv/bin/python -m pytest tests/unit/sdk/ --ignore=tests/unit/sdk/database
============================= 306 passed in 2.11s ==============================
```

(304 → 306 with the new `TestDeepSeekTPAllReduce` class.)

#### Related Issues:

- Relates to internal Kimi-K2.5 TTFT-accuracy work tracked separately. Follow-ups (TRT-LLM widening, SGLang op layout, fused-decode AR) will be filed as separate issues.

### CODE REVIEW

**Review by Harrilee — APPROVED** [review]
A lot gap has been reduced!
